mev_inspect/arbitrages.py

```diff
@@ -23,70 +23,124 @@ def get_arbitrages(swaps: List[Swap]) -> List[Arbitrage]:
 
 
 def _get_arbitrages_from_swaps(swaps: List[Swap]) -> List[Arbitrage]:
-    pool_addresses = {swap.pool_address for swap in swaps}
+    """
+    An arbitrage is defined as multiple swaps in a series that result in the initial token being returned.
 
-    all_arbitrages = []
-
-    for index, first_swap in enumerate(swaps):
-        other_swaps = swaps[:index] + swaps[index + 1 :]
+    Input swap list does not have to be ordered and can have excess swaps but cannot miss swaps in the path of arbitrage.
+    Ex: [WETH -> UNI, UNI -> DAI, DAI -> WETH]
+    Non-working Ex: [WETH -> UNI, DAI -> WETH]
 
-        if first_swap.from_address not in pool_addresses:
-            arbitrage = _get_arbitrage_starting_with_swap(first_swap, other_swaps)
+    Algorithm works by first finding possible start and end swap pairs (strategy defined in _get_all_start_end_swaps) then finding all possible routes between them.
+    """
 
-            if arbitrage is not None:
-                all_arbitrages.append(arbitrage)
+    all_arbitrages = []
 
-    return all_arbitrages
+    start_ends = _get_all_start_end_swaps(swaps)
+    if len(start_ends) == 0:
+        return []
 
+    filtered_start_ends = filter(
```

Comment:
what does this step get us?

I think as long as all of the swaps happened in the same transaction and form a circuit, it shouldn't matter the ordering within the trace

Reply:
Just saw the piece in the PR mentioning it

I would think these
```
one of the intermediate transactions in the swap route does not use all of the output of the previous swap
one of the intermediate transactions in the swap route uses more than the output of the previous swap
```
are still arbs - we might need some better ways to describe what's happening though

Reply:
Yeah so currently we still pick these up. Vast majority of the time this happens is due to one of the following two cases:
- Rounding error in the contract: Sometimes MEV bot contracts only care about the first 12 decimals for example.
- Contract is doing A -> B -> A, happens to have some extra B sitting around when it does the swap and just takes all of its B and puts it in the B->A swap.
From my testing it's usually an insignificant sum (10^-6 in size), so I don't think it's worth worrying about unless we find some seriously mis-evaluated arbs down the road.

## Why the filtering start_ends matters:
If we have A -> B -> C -> A, this comes out to these swaps:
1. A->B
2. B->C
3. C->A

Without any assumptions this can be routed as:
`A->B->C->A`, `B->C->A->B`, `C->A->B->C`
So we need to add some assumptions to find a way to choose the starting and ending point. 

**We can't use `to` and `from` addresses of the swaps** because of the case where the bot contract is executing swaps individually:
- Case where this would work: `BOT->Pool A/B->Pool B/C->Pool A/C->BOT`
- Case where this would break: `BOT->Pool A/B->BOT` / `BOT->Pool B/C->BOT` / `BOT->Pool A/C->BOT`

The next obvious way would be based on transaction ordering. Unfortunately **we can't exclusively rely on this due to flash swaps** (previously referenced [tx](https://etherscan.io/tx/0xfcf4558f6432689ea57737fe63124a5ec39fd6ba6aaf198df13a825dd599bffc), uni flash swap [docs](https://docs.uniswap.org/protocol/V2/concepts/core-concepts/flash-swaps)) which allow the actual swap events and transfer events to be out of order. 

Thus we rely on profitability for detecting the start/end swap. Two notes here
- `mev-inspect-rs` does this as well
- From my experience (plz correct me if i'm wrong) no successful MEV bot contract is dumb enough to execute a series of swaps (and pay the miner) if it is not sum profitable after on-chain execution.

In summary:
- Relying on transaction ordering via `trace_address` isn't perfect
- Relying on profitability of start vs final swap isn't perfect

The current solution is to semi rely on both and it seems to work quite well in practice.
- Rely on profitability for start/end swaps
- Rely on transaction ordering to the extent that swaps at least happen at the same trace level. This allows 
  - (most) out of order flash swaps through the filter 
  - filters out the arbs where an intermediate swap is detected as a start/end swap due to `prev_swap.token_out_amount != next_swap.token_in_amount` from dust discrepancies (details in COMMENT#FIL THIS IN SAM

Reply:
re:
> From my experience (plz correct me if i'm wrong) no successful MEV bot contract is dumb enough to execute a series of swaps (and pay the miner) if it is not sum profitable after on-chain execution.

some bots exclude the check for gas savings...

Reply:
Instead of using profitability or trace ordering to determine where to start, is it possible to just assume the start / end will be whichever address isn't a pool? I think that would work for both cases

Reply:
Ok, fair enough, we can try to include non-profitable. Here's an [ex tx](https://etherscan.io/tx/0x4f2ec102adc25ceb8440a558e3d82e97a4be3ba839dd055db2749ba194d44d8b). 

99+% of arbs I've looked at look like the following two cases:
Case 1 (fully routed. one transfer from the bot, one transfer to the bot):
`Bot -> A/B -> B/C -> C/A -> Bot`

Case 2 (Bot is always intermediary):
`Bot -> A/B -> Bot -> B/C -> Bot -> A/C -> Bot`

Case 1:
- These are easy, we can only calculate one route to and from the bot, making starting ending swap obvious
- (These are also commonly out of order based on trace_address due to flash swaps: [ex](https://etherscan.io/tx/0x0542cbf5ab7ed4b616d01de4bfa4f870a4fc7b1ed7288322fe9cddf0dbd5149b))

Case 2:
- These are the issue, for 3 swaps, we can calculate 3 different starting/ending points and therefore 3 routes
- (In the hundreds of examples I've checked, these are in order: [ex](https://etherscan.io/tx/0x4f2ec102adc25ceb8440a558e3d82e97a4be3ba839dd055db2749ba194d44d8b), but I can imagine a case where they are not)

Proposed solution: Try to fully route a set of swaps into arbs (not relying on profitability). If we only get one valid route, we assume a valid arb (this would cover all of Case 1). If we get multiple valid routes (Case 2) with overlapping swaps, we assume only one is valid and make that decision based on trace_address ordering. Alternatively we could make that decision based on profitability. Let me know what you think!

Reply:
Again I really appreciate you writing these up 🙏 This is super helpful

For case 2, I think the resulting arbitrage object will look the same regardless of which we choose as the start right? As long as we pick a point which isn't a pool address. So first in transaction should be solid.

Reply:
Overall this sounds solid though and nice and simple!

Reply:
Yes no worries!

>For case 2, I think the resulting arbitrage object will look the same regardless of which we choose as the start right?

No because the `profit_token` `profit_amount` and `swaps` order will be different. 

I'll rewrite to the new algorithm and push. Will be up shortly.

Reply:
Updated. Also updated the PR description.